### DISCUSSION TITLE
Does this work on Android?

### DISCUSSION THREAD

**Discussion by yantlord** [Hakky54/ayza#179] [Q&A]
Hello :)

The library looks fantastic, really nice work and thanks for sharing it. This stuff is very confusing, I've been having a hard time getting a TLS-enabled server running. I'm wondering if anyone has tested this on Android (Kotlin with Ktor/Netty)? Specifically using the AndroidKeyStore, I'm also wondering how I would configure things to try it - should I feed the keystore in as a InputStream?

Thanks again :)

**Comment by Hakky54** [comment]
Hi @yantlord 

Thank you for the compliments, I really appreciate it! ❤️ 

SSL Configuration is actually not that complex, but the verbosity of Java makes it in my opinion a bit more complex than needed for the developers. I tried to simplify it with this library so the end-user can easily set it up without the need of understanding the low level configuration. This library has proven to be working on Android and also with Netty, it has even a separate library `io.github.hakky54:sslcontext-kickstart-for-netty:7.4.1` to convert the SSL configuration from `SSLFactory` to a Netty based `SslContext`. See here for an example: 
https://github.com/Hakky54/java-tutorials/blob/7d0cef9b79f8816a8f705dcd57dca4b3f326c43f/instant-server-ssl-reloading-with-netty/netty-server/src/main/java/nl/altindag/server/App.java#L19

I have analysed the source code of Ktor with Netty and it seems like they didn't exposed the ssl configuration for the server through a custom `SslContext`, so it is only possible to configure ssl with the following properties as shown on their documentation page: https://ktor.io/docs/ssl.html#embedded-server and not through this library. But that kind of configuration should do the trick for most of the use cases in my opinion.

If the Ktor developers expose the sslcontext in the server configuration through a setter or as constructor parameter for the configuration, than this library can be used.

**Reply by midthegap** [reply]
Hi all,

I add my compliments for this cool project, and also for all the time spent in supporting it from its author. Really good work!

I am also trying to use the project in Android, but it seems you will need to change some dependancies. Base64 enconding is done via java.util.Base6, unfortunately you don't have this class on Android. You get android.util.Base64 or something like that.

In my project I have OkHttp 4.x, and it seems that documentation on SSL mutual authentication is not up-to-date on the OkHttp site. Thus I don't know if sslcontext-kickstart is working with OkHttp4 or it requires OkHttp3. I am now struggling with no luck.
But this is a question out of scope for sslcontext-kickstart.

Thank you for your work!

**Reply by Hakky54** [reply]
Hi @midthegap 

Thank you mate for your kind words ❤️ 

`java.util.Base64` is missing when you use API Level 25 or older. If your are able to switch to API Level 26 (Android 7.1 Nougat) which was released in 2016 or newer it will work. I found the documentation here: https://developer.android.com/reference/java/util/Base64

If you need a workaround you could try using `io.github.hakky54:sslcontext-kickstart-for-pem` see here for the full documentation of using pem files: https://github.com/Hakky54/sslcontext-kickstart#using-pem-files The pemutils is able to handle all pem files and does not use base64 util.

> In my project I have OkHttp 4.x, and it seems that documentation on SSL mutual authentication is not up-to-date on the OkHttp site. Thus I don't know if sslcontext-kickstart is working with OkHttp4 or it requires OkHttp3. I am now struggling with no luck.
But this is a question out of scope for sslcontext-kickstart.


I was not aware that there was OkHttp4, I need to update my examples 😄 I have here the documentation for OkHttp3 https://sslcontext-kickstart.com/client/okhttp.html and it seems like the ssl configuration is exactly the same as the older version of OkHttp, see here for the latest version stated on their documentation page: https://square.github.io/okhttp/features/https/#customizing-trusted-certificates-kt-java

So I am assuming it will just work, but I have the feeling you have a different issue. Can you retry and share your stacktrace?
